databases/kbase/user_federation.go

```diff
@@ -0,0 +1,289 @@
+	//
+	// * there are 2 entries separated by exactly one comma
+	// * exactly one of the entries is a well-formed ORCID (xxxx-xxxx-xxxx-xxxx)
+	// * the other entry is a non-empty string with no special characters
+	//
+	// Finally, the structure of all the lines in the file must agree. Every line
+	// that doesn't conform to these requirements is ignored. If there's at least
+	// one valid line, we clear the existing KBase user table and add each
+	// (ORCID, user) pair to the user table.
+	orcidColumn := -1
+	userColumn := -1
+	orcidsForUsers := make(map[string]string)
+	usersForOrcids := make(map[string]string)
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+		cells := strings.Split(line, ",")
+		if len(cells) != 2 {
+			return nil, InvalidKBaseUserSpreadsheet{
+				File:    kbaseUserTableFile,
+				Message: fmt.Sprintf("%d comma-separated columns found (2 expected)", len(cells)),
+			}
+		}
+
+		if orcidColumn == -1 { // find the column with an ORCID
+			for i := 0; i < 2; i++ {
+				if isOrcid(cells[i]) {
+					orcidColumn = i
+					userColumn = (i + 1) % 2 // user column's the other one
+				}
+			}
+		} else if !isOrcid(cells[orcidColumn]) {
+			// we've already established the ORCID column, but this line disagrees,
+			// so the whole file is suspect
+			return nil, InvalidKBaseUserSpreadsheet{
+				File:    kbaseUserTableFile,
+				Message: "Different lines list username, ORCID data in different columns",
+			}
+		}
+
+		if orcidColumn != -1 {
+			orcid := cells[orcidColumn]
+			// ORCID column's okay, but what about the user column?
+			if !isUsername(cells[userColumn]) {
+				continue
+			}
+			username := cells[userColumn]
+
+			// have we seen this ORCID or username before? It's okay, as long as everything
+			// is consistent
+			if existingUser, found := usersForOrcids[orcid]; found {
+				if existingUser != orcid {
+					return nil, InvalidKBaseUserSpreadsheet{
+						File:    kbaseUserTableFile,
+						Message: fmt.Sprintf("ORCID %s is associated with multiple users", orcid),
+					}
+				}
+			} else {
+				usersForOrcids[orcid] = username
+			}
+			if existingOrcid, found := orcidsForUsers[username]; found {
+				if existingOrcid != orcid {
+					return nil, InvalidKBaseUserSpreadsheet{
+						File:    kbaseUserTableFile,
+						Message: fmt.Sprintf("User %s has multiple ORCIDs", username),
+					}
+				}
+			} else {
+				orcidsForUsers[username] = orcid
+			}
+		}
+	}
+
+	if len(usersForOrcids) == 0 {
+		return nil, InvalidKBaseUserSpreadsheet{
+			File:    kbaseUserTableFile,
+			Message: "No valid username/ORCID pairs found",
+		}
+	}
+
+	return usersForOrcids, nil
+}
+
+// returns true iff s contains a valid username
+func isUsername(s string) bool {
+	return len(s) > 0 && !strings.ContainsFunc(s, func(c rune) bool {
+		return !unicode.IsLetter(c) && !unicode.IsDigit(c) && c != '_'
+	})
+}
+
+// returns true iff s contains a valid ORCID
```

Comment:
I'd just add the ORCID format here so people don't have to decode the code

nnnn-nnnn-nnnn-nnnn or nnnn-nnnn-nnnn-nnnX